File: chromium/chrome/browser/upboarding/query_tiles/android/tile_provider_factory.cc
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/jni_android.h"
#include "chrome/android/chrome_jni_headers/TileProviderFactory_jni.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_android.h"
#include "chrome/browser/profiles/profile_key.h"
#include "chrome/browser/upboarding/query_tiles/android/tile_provider_bridge.h"
#include "chrome/browser/upboarding/query_tiles/tile_service_factory.h"

// Takes a Java Profile and returns a Java TileProvider.
static base::android::ScopedJavaLocalRef<jobject>
JNI_TileProviderFactory_GetForProfile(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& j_profile) {
  Profile* profile = ProfileAndroid::FromProfileAndroid(j_profile);
  ProfileKey* profile_key = profile->GetProfileKey();

  // Return null if there is no reasonable context for the provided Java
  // profile.
  if (profile_key == nullptr)
    return base::android::ScopedJavaLocalRef<jobject>();

  upboarding::TileService* tile_service =
      upboarding::TileServiceFactory::GetInstance()->GetForKey(profile_key);
  return upboarding::TileProviderBridge::GetBridgeForTileService(tile_service);
}

File: chromium/chrome/android/java/src/org/chromium/chrome/browser/query_tiles/TileProviderFactory.java
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.query_tiles;

import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.profiles.Profile;

/**
 * Basic factory that creates and returns an {@link TileProvider} that is attached
 * natively to the given {@link Profile}.
 */
public class TileProviderFactory {
    private static TileProvider sTileProvider;

    /**
     * Used to get access to the tile provider backend.
     * @return An {@link TileProvider} instance.
     */
    public static TileProvider getForProfile(Profile profile) {
        if (sTileProvider == null) {
            sTileProvider = TileProviderFactoryJni.get().getForProfile(profile);
        }

        return sTileProvider;
    }

    /** For testing only. */
    public static void setTileProviderForTesting(TileProvider provider) {
        sTileProvider = provider;
    }

    @NativeMethods
    interface Natives {
        TileProvider getForProfile(Profile profile);
    }
}
